nanofi/examples/tail_file.c

```diff
@@ -0,0 +1,270 @@
+                transmit_flowfile(flowfiles->records[i], instance);
+            }
+        }
+    }
+}
+
+char * get_string(char * begin, char * end) {
+
+    if (begin == NULL || (end != NULL && end < begin)) {
+        return NULL;
+    }
+
+    int len = 0;
+    char * buff = NULL;
+
+    if (begin == end) {
+        buff = (char *)malloc(2 * sizeof(char));
+        strncpy(buff, begin, 1);
+        buff[1] = '\0';
+        return buff;
+    }
+
+    if (end) {
+        len = end - begin;
+        buff = (char *)malloc((len+1) * sizeof(char));
+        strncpy(buff, begin, len);
+        buff[len] = '\0';
+    }
+    else {
+        len = strlen(begin);
+        buff = (char *)malloc((len+1) * sizeof(char));
+        strncpy(buff, begin, len);
+        buff[len] = '\0';
+    }
+    return buff;
+}
+
+void split_string(char * str, char delim, char ** collector, int * num_strings)
+{
+    if (!str) return;
+
+    while (*str == delim) str++;
+
+    char * p = strchr(str, delim);
+    if (p) {
+        char * s = get_string(str, p);
+        if (s == NULL) {
+            return;
+        }
+        collector[*num_strings] = s;
+        (*num_strings)++;
+        split_string(p+1, delim, collector, num_strings);
+    }
+    else {
+        char * s = get_string(str, NULL);
+        if (s == NULL) {
+            return;
+        }
+        collector[*num_strings] = s;
+        (*num_strings)++;
+    }
+}
+
+void set_offset(int offset) {
+    fileoffset->offset = offset;
+}
+
+int get_offset() {
+    return fileoffset->offset;
+}
+
+void free_all_strings(char ** strings, int num_strings) {
+    for (int i = 0; i < num_strings; ++i) {
+        if (strings[i]) free(strings[i]);
+    }
+}
+
+void on_trigger_callback(processor_session * ps, processor_context * ctx) {
+
+    char file_path[50];
+    char delimiter[2];
+
+    if (get_property(ctx, "file_path", file_path, 50) != 0) {
+        return;
+    }
+
+    if (get_property(ctx, "delimiter", delimiter, 2) != 0) {
+        return;
+    }
+
+    char delim = '\0';
+    if (strlen(delimiter) > 0) {
+        delim = delimiter[0];
+    }
+
+    if (delim == '\0') {
+        printf("Invalid delimiter \n");
+        return;
+    }
+
+    if (delim == '\\') {
+      if (strlen(delimiter) > 1) {
+        switch (delimiter[1]) {
+          case 'r':
+            delim = '\r';
+            break;
+          case 't':
+            delim = '\t';
+            break;
+          case 'n':
+            delim = '\n';
+            break;
+          case '\\':
+            delim = '\\';
+            break;
+          default:
+            break;
+        }
+      }
+    }
+
+    int curr_offset = get_offset();
+    int max_bytes_read = 4096;
+    char buff[max_bytes_read];
+    FILE * fp = fopen(file_path, "rb");
+    if (!fp) return;
+    fseek(fp, curr_offset, SEEK_SET);
+    while (fgets(buff, max_bytes_read, fp) != NULL) {
+        set_offset(ftell(fp));
+        char * strings[max_bytes_read];
```

Comment:
The number of delimiters is usually much smaller than the number of characters in the stream.
Preallocating as many possible substrings as there are characters is wasteful.
I would prefer for split_string to have a different contract, see my comments there.

Reply:
Updated code.